rholang/src/test/scala/coop/rchain/rholang/interpreter/RegistrySpec.scala

```diff
@@ -0,0 +1,2403 @@
+package coop.rchain.rholang.interpreter
+
+import com.google.protobuf.ByteString
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.models.Channel.ChannelInstance.{ChanVar, Quote}
+import coop.rchain.models.Expr.ExprInstance._
+import coop.rchain.models._
+import coop.rchain.models.rholang.implicits._
+import coop.rchain.rholang.interpreter.accounting.{CostAccount, CostAccountingAlg}
+import coop.rchain.rholang.interpreter.storage.implicits._
+import coop.rchain.rspace._
+import coop.rchain.rspace.internal.{Datum, Row, WaitingContinuation}
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.{FlatSpec, Matchers}
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+trait RegistryTester extends PersistentStoreTester {
+  implicit val errorLog = new ErrorLog()
+  implicit val costAccounting =
+    CostAccountingAlg.unsafe[Task](CostAccount.zero)
+  def withRegistryAndTestSpace[R](
+      f: (Reduce[Task], 
+          ISpace[Channel,
+                 BindPattern,
+                 ListChannelWithRandom,
+                 ListChannelWithRandom,
+                 TaggedContinuation]) => R
+          ): R = {
+    withTestSpace { space =>
+      lazy val registry: Registry = new Registry(space, dispatcher)
+      lazy val dispatcher: Dispatch[Task, ListChannelWithRandom, TaggedContinuation] =
+        RholangAndScalaDispatcher.create[Task, Task.Par](space, registry.testingDispatchTable, Map.empty)
```

Comment:
If you create a `dispatcher` with `urnMap` 
```scala
lazy val urnMap: Map[String, Par] = Map(
  "rho:registry:lookup" -> GPrivate(ByteString.copyFrom(Array[Byte](10))),
  "rho:registry:insert" -> GPrivate(ByteString.copyFrom(Array[Byte](12))),
  "rho:registry:delete" -> GPrivate(ByteString.copyFrom(Array[Byte](14)))
)
lazy val dispatcher: Dispatch[Task, ListChannelWithRandom, TaggedContinuation] =
  RholangAndScalaDispatcher.create[Task, Task.Par](space, registry.testingDispatchTable, urnMap)
```

then now instead of hardcoding ASCII we can parse rholang code:
```scala
def parse(rho: String): Par = {
  val source     = s"for (@x3, @x2, @x1, @x0 <- @0) { $rho }"
  val astTerm = Interpreter.buildNormalizedTerm(new StringReader(rho)).value
  val rhoTerm = astTerm.receives.head.body
  rhoTerm
}
…
val rholangString = """new r(`rho:registry:lookup`) in {
        #        r!("0897".hexToBytes(), "result0") |
        #        r!("0897dcbcb195efbad3b7ca343f50713e".hexToBytes(), "result1") |
        #        r!("0897e775426a859c8d3b25df4508cda9".hexToBytes(), "result2")
        #      }""".stripMargin('#')
val lookupRholang = parse(rholangString)
…
val result = withRegistryAndTestSpace { (reducer, space) =>
      implicit val env    = Env[Par]()
      Console.println(StoragePrinter.prettyPrint(space.store))
      val resultTask = for {
        _ <- reducer.eval(Par().addSends(rootSend, branchSend))
        _ <- reducer.eval(lookupRholang)
      } yield space.store.toMap

      Await.result(resultTask.runAsync, 3.seconds)
    }
```

Reply:
Also, doesn't it show that a malicious validator can modify the code of rchain so that any data sent to `rho:registry:*` is also sent to the name he controls?